libkbfs/journal_block_server_test.go

```diff
@@ -15,24 +15,29 @@ import (
 )
 
 func setupJournalBlockServerTest(t *testing.T) (
-	tempdir string, config Config, jServer *JournalServer) {
+	tempdir string, config *ConfigLocal, jServer *JournalServer) {
+	config = MakeTestConfigOrBust(t, "test_user")
+
 	tempdir, err := ioutil.TempDir(os.TempDir(), "journal_block_server")
 	require.NoError(t, err)
-	config = MakeTestConfigOrBust(t, "test_user")
-	log := config.MakeLogger("")
-	jServer = makeJournalServer(
-		config, log, tempdir, config.BlockCache(), config.DirtyBlockCache(),
-		config.BlockServer(), config.MDOps(), nil, nil)
-	ctx := context.Background()
-	err = jServer.EnableExistingJournals(
-		ctx, TLFJournalBackgroundWorkPaused)
+	// Clean up the tempdir if anything in the setup fails/panics.
+	defer func() {
+		if r := recover(); r != nil {
+			CheckConfigAndShutdown(t, config)
```

Comment:
I dug into this a bit more, and it turns out I was using `recover()` incorrectly, since I wasn't re-panicking with the `recover` value. Also, `t.FailNow` doesn't actually panic, but calls `Goexit()`, which makes `recover()` return `nil`, which is indistinguishable from normal execution. Urgh.

So I fixed up all the setup/teardown calls to do the following:
1) Only use `assert` and not `require` in `teardown`
2) Have a `setupSucceeded` flag in `setup` and check it when deciding whether or not to clean up on exit.
3) Have all the cleanup functions also use `assert` and not `require`.

So sometimes the cleanup code in `setup` shares code with `teardown`, but they differ enough that I don't think it's worth complicating `teardown` to handle partial setup states. In this case, `setupJournalBlockServerTest` now has two `defer` blocks to clean up `tempdir` and `config`, and `teardownJournalBlockServerTest` has similar code. I guess I could make the latter check whether `config` is `nil`, but I don't know if it's that much better. I can change it if you feel strongly.